### PR TITLE
Clean up tuple elements after unreachable

### PR DESCRIPTION
**Pull request by MakeNowJust**
Resolve https://github.com/crystal-lang/crystal/pull/6628#issuecomment-417981480 problem.

Small reproducible example:

```crystal
def foo
  yield({1, while true; end})
end

foo { |a, b| }
```

Result:

```console
$ crystal run foo.cr
Missing hash key: "b" (KeyError)
  from Hash(String, Crystal::CodeGenVisitor::LLVMVar)+@Hash(K, V)#[]<String>:Crystal::CodeGenVisitor::LLVMVar
  from Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil
  from Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool
  from Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil
  from Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil
  from Crystal::Compiler#codegen<Crystal::Program, Crystal::ASTNode+, Array(Crystal::Compiler::Source), String>:(Tuple(Array(Crystal::Compiler::CompilationUnit), Array(String)) | Nil)
  from Crystal::Compiler#compile<Array(Crystal::Compiler::Source), String>:Crystal::Compiler::Result
  from Crystal::Command#run_command<Bool>:Nil
  from Crystal::Command#run:(Bool | Crystal::Compiler::Result | Nil)
  from main
```

This reason is `yield` tuple destructing against the tuple having unreachable elements. An unreachable element should be treated by `CleanupTransformer`. This PR adds cleanup process for `TupleLiteral` and `NamedTupleLiteral` to `CleanupTransformer`.

Thank you.

### CODE REVIEW

**Comment by bew** [comment]
Sorry but I don't understand the problem and the solution here..

From #6628 I think you're using `while true; end` to get a `NoReturn` type, right? Can you add a note about this?

Also it starts as a `Tuple` and you transform it to an `Expressions` right? How come this is not a problem?? You're changing the type arbitrarily, what if I'm using the result as a `Tuple`??

(I don't understand the problem and the change..)

**Comment by asterite** [comment]
Yeah, the fix doesn't seem to be right...

**Comment by MakeNowJust** [comment]
`while true; end` is simple way to get `NoReturn` expression. (@asterite please tell me if you have simpler way.)

We cannot get the tuple having `NoReturn` element, so we can eliminate elements after unreachable. In fact, this kind optimization is applied to all other cases by `CleanupTransformer`. For example, I have this hash literal:

```crystal
{:foo => 1, :bar => while true; end, :baz => 2}
```

Finally, through `CleanupTransformer`, it is transformed to:

```crystal
__temp_xxx = ::Hash(typeof(:foo, :bar, :baz), typeof(1, while true; end, 2)).new
__temp_xxx[:foo] = 1
__temp_xxx
:bar
while true; end
```

I think the reason of https://github.com/crystal-lang/crystal/pull/6628#issuecomment-417981480 is missing to apply this optimization to `TupleLiteral` and it affects the codegen phase as a result.

File: spec/compiler/codegen/def_spec.cr [hunk 1]
```diff
@@ -558,4 +558,14 @@ describe "Code gen: def" do
       foo(1 || true)
       ))
   end
+
+  it "codegens yield with destructing tuple having unrachable element" do
```
Comment (r00ster91) [inline spec/compiler/codegen/def_spec.cr]: Typo: unrachable

**Comment by MakeNowJust** [comment]
@asterite Could you review it with seriousness? I think it is the only true way to fix https://github.com/crystal-lang/crystal/pull/6628#issuecomment-417981480.

Do you expect we can fix it by avoiding `KeyError` with using `[]?`, like:

```diff
diff --git a/src/compiler/crystal/codegen/codegen.cr b/src/compiler/crystal/codegen/codegen.cr
index 490abfe77..6ddca53ae 100644
--- a/src/compiler/crystal/codegen/codegen.cr
+++ b/src/compiler/crystal/codegen/codegen.cr
@@ -1437,15 +1437,17 @@ module Crystal
               if arg && arg.name != "_"
                 t_type = tuple_type
                 t_value = codegen_tuple_indexer(exp_type, exp_value, i)
-                block_var = block_context.vars[arg.name]
-                assign block_var.pointer, block_var.type, t_type, t_value
+                if block_var = block_context.vars[arg.name]?
+                  assign block_var.pointer, block_var.type, t_type, t_value
+                end
               end
             end
           else
             exp_values.each_with_index do |(exp_value, exp_type), i|
               if (arg = block.args[i]?) && arg.name != "_"
-                block_var = block_context.vars[arg.name]
-                assign block_var.pointer, block_var.type, exp_type, exp_value
+                if block_var = block_context.vars[arg.name]?
+                  assign block_var.pointer, block_var.type, exp_type, exp_value
+                end
               end
             end
           end
```

But the above patch does not resolve anything.

```console
$ cat no-return-hash.cr
p({} of NoReturn => NoReturn)
$ bin/crystal run no-return-hash.cr
Using compiled compiler at `.build/crystal'
Error in src/hash.cr:857: BUG: missing context var: key

        key.inspect(io)
        ^~~
```

The true problem is `CleanupTransformer`'s bug, that is missing to transform `TupleLiteral` with `NoReturn` element. The codegen phase expects the program eliminated expressions after `NoReturn` expression. When `CleanupTransform` is not perfect, it makes sense that problems happen on the codegen phase.

**Comment by asterite** [comment]
Now that I see it again it's almost OK. What happens if it's a tuple with one element that's NoReturn? I think it'll be a compiler bug with this code.

**Pull request merged**